### PR TITLE
[WIP] Respect cursor_start and cursor_end in company

### LINKED ISSUE #436: Autocomplete quirks with company backend
Hi @millejoh, thanks for your work on this amazing package. 
I'm just setting up EIN in my Emacs and I'm experiencing some undesired behaviour.

Let's say I have imported a module in one cell
```
from sklearn import model_selection
```

If in another cell I try to invoke autocomplete after the dot
```
model_selection.
```

the typed expression gets replaced by the selected suggestion (*train_test_split* in this case)  instead of being appended. 

EXPECTED
```
model_selection.train_test_split
```

ACTUAL
```
train_test_split
```

Here's my configuration 
```
(use-package ein
  :pin melpa
  :config
  (setq ein:use-auto-complete t)
  (setq ein:complete-on-dot t)
  (setq ein:completion-backend 'ein:use-company-backend)
  (setq ein:use-auto-complete-superpack nil)
  (setq ein:use-smartrep nil)
)
```
I tried to look in logs in the debug mode, but haven't found anything that seemed useful. 
Using EIN *ein-20181213.1844*.

### PR DESCRIPTION
**Pull request by dickmao**
With respect to `ein-company`,
Py2 kernels prepend prefixes to completions, e.g., np.array.
Py3 kernels do not, e.g., array.
The `:cursor_start` and `:cursor_end` fields from ipykernel's `complete_reply` tell us whether we need to
prepend the prefix ourselves.

Fixes #436

### CODE REVIEW

**Comment by millejoh** [comment]
Thank you. Where did you find the documentation showing the difference between ipy2 and ipy3 kernels? The documentation I looked at gave the impression that candidates returned in `complete_reply` messages has a more consisten format. I guess `ein:completions--find-cached-completion` is a bit wordy, but thought it explained pretty well what was going on. Do you mind sharing your rationale for changing the name?

**Comment by dickmao** [comment]
Sorry, can you revert? [WIP] means "I am in the middle of addressing this concern, and if you have other ideas maybe we can congregate here"  Until [WIP] is removed, it generally isn't ready for merging to master.  Thanks and sorry for the runaround.

**Comment by millejoh** [comment]
No problem - reverted. Sorry about the miscommunication.

**Comment by dickmao** [comment]
My findings regarding ipy2 and ipy3 stem directly from the evidence provided by xelibrion in #436.  I was able to successfully reproduce the misbehaviour.

`ein:completions--find-cached-completion` should not have two dashes because it is called outside of its source file `ein-completer.el`.  Ideally, it should be prefixed `ein:completer` instead of `ein:completions`.  `ein:completer-find-cached-completion` would be a fine name, but it happened to take lengthy arguments that caused the line to exceed 80 characters.  My brain has difficulty parsing wrapped lines and function calls that need to get broken up over several lines.  

`ein:completions-get-cached` saved me about 15 characters allowing me to fit the call on a single line and conveys most of the meaning.

**Comment by millejoh** [comment]
Thanks - I didn't give consideration to line length; something I will try to do in the future. 

I'm also not sure this is so much a python 3 issue as it is with IPython 7.2 - completion was working fine for until I upgraded my installation.

**Comment by dickmao** [comment]
> I'm also not sure this is so much a python 3 issue as it is with IPython 7.2 - completion was working fine for until I upgraded my installation.

Yes, that is correct.
